lib/ops/opsservice/shrink.go

```diff
@@ -454,62 +455,67 @@ func (s *site) waitForServerToDisappear(hostname string) error {
 	return trace.Wrap(err)
 }
 
-func (s *site) removeFromEtcd(ctx *operationContext, runner *serverRunner, server storage.Server) error {
-	out, err := runner.Run(s.etcdctlCommand("member", "list")...)
-	if err != nil {
-		return trace.Wrap(err, "failed to list etcd members: %s", out)
-	}
-
-	provisionedServer := ProvisionedServer{Server: server}
-	memberID, err := utils.FindETCDMemberID(
-		string(out), provisionedServer.EtcdMemberName(s.domainName))
-	if err != nil {
+func (s *site) removeFromEtcd(ctx context.Context, opCtx *operationContext, server storage.Server) error {
+	peerURL := server.EtcdPeerURL()
+	logger := opCtx.WithField("peer", peerURL)
+	logger.Info("Remove peer from etcd cluster.")
+	b := utils.NewExponentialBackOff(defaults.EtcdRemoveMemberTimeout)
+	return utils.RetryTransient(ctx, b, func() error {
+		client, err := clients.DefaultEtcdMembers()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		members, err := client.List(ctx)
+		logger.WithError(err).WithField("peers", members).Info("Etcd members.")
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		member := utils.EtcdHasMember(members, peerURL)
+		if member == nil {
+			logger.Info("Peer not found.")
+			return nil
+		}
+		err = client.Remove(ctx, member.ID)
+		logger.WithError(err).Info("Removed etcd peer.")
 		return trace.Wrap(err)
-	}
-
-	ctx.Debugf("found member: %v", memberID)
-
-	out, err = runner.Run(s.etcdctlCommand("member", "remove", memberID)...)
-	if err != nil {
-		return trace.Wrap(err, "failed to remove etcd member: %s", out)
-	}
-
-	return nil
+	})
 }
 
 func (s *site) uninstallSystem(ctx *operationContext, runner *serverRunner) error {
 	commands := [][]string{
-		s.gravityCommand("system", "uninstall", "--confirm"),
+		s.gravityCommand("system", "uninstall", "--confirm", "--no-uninstall-service"),
```

Comment:
Hidden flag to avoid uninstalling the agent service as long as the shrink operation is not complete as `system uninstall` automatically uninstalls and removes the agent services.

Reply:
Is the service really still needed after "system uninstall" completes? Why not just let it uninstall itself (as a very last step of system uninstall)?

Reply:
Not needed. I've revised the uninstall path and made the agent cleanup the last step. Also made the uninstall step fail if a package service fails to uninstall since removing state at this point will render further attempts to uninstall the services impossible (i.e. if the planet service fails for some reason - it would be impossible to clean up manually if we proceeded and removed the rootfs directory).